core/src/main/java/org/elasticsearch/action/support/replication/ReplicationOperation.java

```diff
@@ -102,27 +110,52 @@ public void execute() throws Exception {
         }
 
         totalShards.incrementAndGet();
-        pendingShards.incrementAndGet();
+        pendingActions.incrementAndGet();
         primaryResult = primary.perform(request);
         final ReplicaRequest replicaRequest = primaryResult.replicaRequest();
         assert replicaRequest.primaryTerm() > 0 : "replicaRequest doesn't have a primary term";
         if (logger.isTraceEnabled()) {
             logger.trace("[{}] op [{}] completed on primary for request [{}]", primaryId, opType, request);
         }
 
-        performOnReplicas(primaryId, replicaRequest);
+        // we have to get a new state after successfully indexing into the primary in order to honour recovery semantics.
+        // we have to make sure that every operation indexed into the primary after recovery start will also be replicated
+        // to the recovery target. If we use an old cluster state, we may miss a relocation that has started since then.
+        ClusterState clusterState = clusterStateSupplier.get();
```

Comment:
Do we have to get this before retrieving the shards? If not, can we move it inside the `if` so we don't do it if we aren't marking unavailable shards as stale